Add support for dual loading

lbulej opened the pull request:
Allows loading the agent library either as a JVM agent or as a normal native library.

Includes several refactorings that "just happened". These include changes to make initialization more straightforward, but also changes that split off bits from the single `ubench.h` include file.

This is supposed to be merged on top of #11.

lbulej committed:
Trigger agent library loading in classes with native methods

The library will only be loaded if it was not loaded as an agent.

lbulej committed:
Introduce jvmutil.c to provide JNI/JVMTI related utility functions

This provides support for working with multile JVMTI contexts.

lbulej committed:
Use separate JVMTI context in counter.c

lbulej committed:
Refactor Java/native thread id mapping

lbulej committed:
Update ubench_benchmark_init signature

lbulej committed:
Setup JVMTI-based staged initialization in threads.c

lbulej committed:
Make ubench_get_current_thread_native_id static

lbulej committed:
Provide JVMTI agent and JNI library points

lbulej committed:
Add support for registering the current Java thread

lbulej committed:
Split off debug/warning/error/fatal printf from ubench.h

This allows using it in other modules without including the whole
ubench.h. Also use other variants in place of fprintf(stderr, ...).

lbulej committed:
Call ubench_event_init from ubench.c

lbulej committed:
Rename ubench_benchmark_init to ubench_measurement_init

This reflects the actual module name.

lbulej committed:
Use specific includes where necessary to avoid pulling everything through ubench.h

lbulej committed:
Simplify dual-loading support

Indicating that the agent has been already loaded by setting a field
in the 'UbenchAgent' class can fail in an environment with multiple
class loaders, because we may fail to find the 'UbenchAgent' class.

Instead, we just check if we can call the agent's native functions and
if not, we try to load it explicitly using the 'loadLibrary()' method.

lbulej committed:
Add struct tags for better warning messages from MSVC

lbulej committed:
Split-off jvmutil.h from ubench.h

lbulej committed:
Reoder fields in jvmti_context_t to avoid MSVC bug

MSVC (at least in version 19.0, i.e., VC++ 2015) mishandles the
jvmtiCapabilities bitfield when using aggregate initializers. This
manifests so that MSVC consider the bitfield storage to be smaller
and puts initializer data into the trailing bytes of the bitfield.

lbulej committed:
Disable specific MSVC warnings

lbulej committed:
Merge branch 'master'

vhotspur reviewed (changes requested):
Overall looks very nice, thank you!

Do you think it would be possible to force the tests to use the new method and thus test both (unless it is already there and I simply do not see it)? Because otherwise I am afraid it will break without us noticing until it is too late :-)

lbulej commented on src/c/jvmutil.c:
```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2014 Charles University in Prague
+ * Copyright 2014 Vojtech Horky
```
> Is this really my code? :-)

Ooops, you have seen straight through my ruse to get people to blame you for my code :-)
Is it OK to leave just CUNI in there?

lbulej replied:
Done.

lbulej commented on src/c/ubench.c:
```diff
@@ -16,44 +16,139 @@
+}
+
+JNIEXPORT void JNICALL
+Agent_OnUnload(JavaVM* vm) {
```
Good point, I was thinking about it while I was away. We could either run test using both methods, or run a just single test using the other method. Any preferences?

lbulej replied:
Any preferences on how would you like it to be integrated?

- One option is to modify the `test` target in `build.xml` so that it would use different loading method for some of the tests (I could not attach a diff, so I'm including it inline). Alternatively, I could modify the `list-events` (or `demo`) target to use `-Djava.library.path` instead of `-agentpath`. All those targets are used in the `gate` target, so modifying any of them should do.

```diff
--- a/build.xml
+++ b/build.xml
@@ -685,53 +685,55 @@
 				<pathelement path="${hamcrest.jar.path}" />
 			</classpath>
 			<compilerarg value="-Xlint:all"/>
 		</javac>
 	</target>
 
 	<macrodef name="self-test">
 		<attribute name="classname" default="mixed" />
+		<attribute name="loadarg" default="-agentpath:${agent.path}" />
 		<attribute name="jvmarg" default="-Xmixed" />
 		<attribute name="progarg" default="mixed" />
 
 		<sequential>
-			<echo level="info" message="[run] java -agentpath:${agent.path} @{jvmarg} @{classname} @{progarg}" />
+			<echo level="info" message="[run] java @{loadarg} @{jvmarg} @{classname} @{progarg}" />
 			<java
 				classname="@{classname}"
 				fork="true"
 				failonerror="true"
 			>
 				<classpath>
 					<pathelement path="${classes.build.dir}" />
 					<pathelement path="${test.classes.build.dir}" />
 				</classpath>
-				<jvmarg value="-agentpath:${agent.path}" />
+				<jvmarg value="@{loadarg}" />
 				<jvmarg value="@{jvmarg}" />
 				<arg value="@{progarg}" />
 			</java>
 		</sequential>
 	</macrodef>
 
 	<target name="test"
 		depends="compile,compile-test"
 		description="Run simple self-tests (obsoleted)."
 	>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.CompilationCounterTest"
 			jvmarg="-Xmixed"
 			progarg="mixed"
 		/>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.CompilationCounterTest"
 			jvmarg="-Xint"
 			progarg="int"
 		/>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.BenchmarkTest"
+			loadarg="-Djava.library.path=${agent.build.dir}"
 		/>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.ThreadTimeTest"
 			progarg="4"
 		/>
 		<self-test
 			classname="cz.cuni.mff.d3s.perf.MeasurementSamplingTest"
 		/>
@@ -808,17 +810,17 @@
 			classname="cz.cuni.mff.d3s.perf.demo.ListEvents"
 			fork="true"
 			failonerror="true"
 		>
 			<classpath>
 				<pathelement path="${classes.build.dir}" />
 				<pathelement path="${demo.classes.build.dir}" />
 			</classpath>
-			<jvmarg value="-agentpath:${agent.path}" />
+			<jvmarg value="-Djava.library.path=${agent.build.dir}" />
 		</java>
 	</target>
 
 	<target name="dist"
 		description="Create distributable ZIP with all sources."
 	>
 		<mkdir dir="${build.dir}" />
 		<zip destfile="${build.dir}/${dist.name}.zip">
```

- Another option is to modify the `self-test` macro to run each test twice, one as agent and one as native library. This would avoid duplicating the information from the `test` target, but the `test` target would run twice as long.
- Finally, I could create a separate target, e.g., `test-loading`, that would test different methods on one test, e.g., which will result in some duplication of entries found in the `test` or `list-events` or `demo` targets.

lbulej replied:
Another option: use the agent in `test` target and the `loadLibrary()` approach in `test-junit` target.

This required a fix both in the `NativeThreads` tests and the threads module initialization, so it actually flushed out some problems, which is good. The change in `build.xml` is then minimal.

lbulej replied:
OK, now I just need to find out why it fails on MacOS. A platform that causes even more problems than Windows should be banned.

vhotspur replied:
I would run it both styles in `gate` (i.e. run both JUnit tests and self-tests in both versions). I hope that we can pass some extra attribute to the target to specify this.

lbulej replied:
> I would run it both styles in gate (i.e. run both JUnit tests and self-tests in both versions). I hope that we can pass some extra attribute to the target to specify this.

Done. All targets that use the agent are now run both ways in the `gate` target.

vhotspur replied:
Thanks!

lbulej committed:
Fix copyright notice in jvmutil

lbulej committed:
Fix the NativeThreadsTest to support non-agent mode

Specifically, this requires the impossibleToRegisterThreadTwice to use
the worker thread instead of the main thread.

lbulej committed:
Update debug messages in ubench.c

lbulej committed:
Extend thread map debugging

lbulej committed:
Add debug prints to NativeThreads native methods

lbulej committed:
Rework thread module initialization

When loaded through 'loadLibrary()', we will not get the 'VM Init'
event to enable automatic thread registration. But we will have a
working JNI, so we can enable the thread lifecycle JVMTI events (if
JVMTI is available).

lbulej committed:
Run junit tests using non-agent mode

lbulej committed:
Avoid checkstyle violations

lbulej committed:
Use library name compatible with loadLibrary() on Mac OS X

lbulej committed:
Test both loading methods in the 'gate' target

lbulej commented:
The windows tests are somewhat fragile, e.g., I saw a failure in a test that measures the thread time of a busy thread and got lower value than expected (7 instead of 8 seconds), and another case where the JVM did not compile anything in mixed mode. I currently don't see anything in the code that should cause it and both failures have a tendency to disappear when rerunning the tests :shrug:

vhotspur commented:
> The windows tests are somewhat fragile, e.g., I saw a failure in a test that measures the thread time of a busy thread and got lower value than expected (7 instead of 8 seconds), and 

Probably we can relax the tests because the CI machines can behave weirdly (might there be an issue in the underlying clock even?).

> another case where the JVM did not compile anything in mixed mode. I currently don't see anything in the code that should cause it and both failures have a tendency to disappear when rerunning the tests shrug

This sounds strange but it happened spuriously to me too.

Probably safe to ignore both.

vhotspur reviewed (approved):
Thanks for the gate update. This looks good to me. Thank you!
